### PR TITLE
Migration update button component

### LINKED ISSUE #10161: Create `Button` migrations
- [x] rename `primary` prop to `variant="primary"`
- [x] rename `plain` prop to `variant="plain"`
- [x] rename `primary` `plain` props to `variant="tertiary"`
- [x] rename `plain` `monochrome` to `variant="monochromePlain"`
- [x] rename `destructive` prop to `tone="critical"` `variant="primary"`
- [x] rename `outline` `destructive` props to `tone="critical"`
- [x] remove `outline` prop
- [x] remove `outline` `monochrome` prop
- [x] remove `monochrome` prop
- [x] handle props in object in `buttonFrom` function
- [ ] Update migration guide

### PR DESCRIPTION
**Pull request by aveline**
### WHY are these changes introduced?

Fixes https://github.com/Shopify/polaris/issues/10161

### WHAT is this pull request doing?

Migration updates `920` files in `web`
Adds comments to `1650` files to be updated manually

### How to 🎩

🖥 [Local development instructions](https://github.com/Shopify/polaris/blob/main/README.md#local-development)
🗒 [General tophatting guidelines](https://github.com/Shopify/polaris/blob/main/documentation/Tophatting.md)
📄 [Changelog guidelines](https://github.com/Shopify/polaris/blob/main/.github/CONTRIBUTING.md#changelog)

<details>
<summary>Copy-paste this code in <code>playground/Playground.tsx</code>:</summary>

```jsx
import React from 'react';
import {Page} from '../src';

export function Playground() {
  return (
    <Page title="Playground">
      {/* Add the code you want to test in here */}
    </Page>
  );
}
```

</details>

### 🎩 checklist

- [ ] Tested on [mobile](https://github.com/Shopify/polaris/blob/main/documentation/Tophatting.md#cross-browser-testing)
- [ ] Tested on [multiple browsers](https://help.shopify.com/en/manual/shopify-admin/supported-browsers)
- [ ] Tested for [accessibility](https://github.com/Shopify/polaris/blob/main/documentation/Accessibility%20testing.md)
- [ ] Updated the component's `README.md` with documentation changes
- [ ] [Tophatted documentation](https://github.com/Shopify/polaris/blob/main/documentation/Tophatting%20documentation.md) changes in the style guide

### CODE REVIEW

**Review by kyledurand — APPROVED** [review]
This is looking great 🙌

File: .changeset/neat-balloons-shop.md [hunk 1]
```diff
@@ -0,0 +1,5 @@
+---
+'@shopify/polaris-migrator': patch
```
Comment (aaronccasanova) [inline .changeset/neat-balloons-shop.md]:
```suggestion
'@shopify/polaris-migrator': minor
```

File: polaris-migrator/src/migrations/v12-react-update-button-components/transform.ts
Comment (aaronccasanova) [inline polaris-migrator/src/migrations/v12-react-update-button-components/transform.ts:1]: Can we rename this migration to `v12-react-update-button-component`? (singular)

File: polaris-migrator/src/migrations/v12-react-update-button-components/transform.ts [hunk 2]
```diff
@@ -0,0 +1,290 @@
...
+) {
+  const source = j(fileInfo.source);
+
+  // If `Button` component name is not imported, exit
+  if (
+    !hasImportDeclaration(j, source, '@shopify/polaris') &&
+    !hasImportSpecifier(j, source, 'Button', '@shopify/polaris')
+  ) {
+    return fileInfo.source;
+  }
```
Comment (aaronccasanova) [inline polaris-migrator/src/migrations/v12-react-update-button-components/transform.ts]: Looks like this conditional will run the migration if there is a `@shopify/polaris` import declaration but no `Button` import specifier.

image

I think we need:

```ts
  // If `Button` component name is not imported, exit
  if (
    !(hasImportDeclaration(j, source, '@shopify/polaris') &&
    hasImportSpecifier(j, source, 'Button', '@shopify/polaris'))
  ) {
    return fileInfo.source;
  }
```

e.g.

image

File: polaris-migrator/src/migrations/v12-react-update-button-components/transform.ts [hunk 3]
```diff
@@ -0,0 +1,290 @@
...
+  const localElementName =
+    getImportSpecifierName(j, source, 'Button', '@shopify/polaris') || 'Button';
+
+  // For each instance of a `Button` get all the attributes and update accordingly
+  source.findJSXElements(localElementName).forEach((element: ASTPath<any>) => {
+    const openingElement = j(element).find(j.JSXOpeningElement).get()
+      .value as JSXOpeningElement;
+
+    const allAttributes = openingElement.attributes ?? [];
+    const jsxAttributes = allAttributes as JSXAttribute[];
```
Comment (aaronccasanova) [inline polaris-migrator/src/migrations/v12-react-update-button-components/transform.ts]: This is missing logic to catch spread syntax:

https://github.com/Shopify/polaris/blob/47652f7d67f98d7bef8ef8485fae803b9f3c1056/polaris-migrator/src/migrations/v10-react-replace-text-components/steps/replace-display-text.ts#L85-L89

**Review by aaronccasanova — APPROVED** [review]

**Pull request merged**